### DISCUSSION TITLE
how to debug why normal mode hangs at  ipxe initializing devices (debug print enclosed)

### DISCUSSION THREAD

**Discussion by kocoman1** [ipxe/ipxe#881] [Q&A]
normal mode hangs at  ipxe initializing devices hang
--
 
debug code added
photo to ascii via word lens app

```
HANDLE PciRoot (0x0) /Pci (0x1C, 0x1) /Pci (0x0, 0x0) /MAC (xxxx, 0x0) DevicePath supported
HANDLE PciRoot (0x0) /Pci (0x1C, 0x1) /Pci (0x0, 0x0) /MAC (xxx. 0x0) DevicePath opened 128x (H) by DxeCore for ‹NULL›
HANDLE Pci Root (0x0) /Pci (0x1C, 0x1) /Pci (0x0, 0x0) /MAC (xxx, 0x0) DevicePath opened 47x (T) by Broadcom 802.11 Wireless Controller Driver
HANDLE PciRoot (0x0) /Pci (0x1C. 0x1) /Pci (0x0, 0x0) /MAC (xxx, 0x0) DevicePath opened 47x (G) by PS/2 Mouse Driver
HANDLE PciRoot (0x0) /Pci (0x1C, 0x1) /Pci (0x0, 0x0) /MAC (x, 0x0) DevicePath opened 47x (T) by Platform Console Management Driver
HANDLE PciRoot (0x0) /Pci (0x1C. 0x1) /Pci (0x0, 0x0) /MAC (xxx, 0x0) DevicePath opened 47x (G) by PS/2 Keyboard Driver
HANDLE PciRoot (0x0) /Pci (0x1C, 0x1) /Pci (0x0, 0x0) /MAC (x, 0x0) Dev icePath opened 47x (T) by Platform Console Management Driver
HANDLE PciRoot (0x0) /Pci (0x1C, 0x1) /Pci (0x0, 0x0) /MAC (x, 0x0) DevicePath opened 15x (T) by Simple Network Protocol Driver
HANDLE Pci Root (0x0) /Pci (0x1C. 0x1) /Pci (0x0, 0x0) /MAC (xxx, 0x0) Nii31 supported
HANDLE Pci Root (0x0) /Pci (0x1C, 0x1) /Pci (0x0, 0x0) /MAC (x, 0x0) Nii31 opened 1x (T) by ipxe.efi
EFIDRV Pci Root (0x0) /Pci (0x1C, 0x1) /Pci (0x0, 0x0) /MAC (x, 0x0) disconnecting existing drivers
EFIDRV PciRoot (0x0) /Pci (0x1C, 0x1) /Pci (0x0, 0x0) /MAC (xxx, 0x0) after disconnecting:
HANDLE PciRoot (0x0) /Pci (0x1C, 0x1) /Pci (0x0, 0x0) /MAC (xxx, 0x0) bf71d7e3-4ae8-47f1-8605-9d576aab4478 supported
HANDLE PciRoot (0x0) /Pci (0x1C, 0x1) /Pci (0x0, 0x0) /MAC (xx, 0x0) 71b4903c-14ec-42c4-bdc6-ce1449930e49 supported
HANDLE PciRoot (0x0) /Pci (0x1C, 0x1) /Pci (0x0, 0x0) /MAC (x  , 0x0) DevicePath supported
HANDLE PciRoot (0x0) /Pci (0x1C, 0x1) /Pci (0x0, 0x0) /MAC (x, 0x0) DevicePath opened 128x (H) by DxeCore for ‹NULL›
HANDLE PciRoot (0x0) /Pci (0x1C, 0x1) /Pci (0x0, 0x0) /MAC (x  , 0x0) Dev icePath opened 47x (T) by Broadcom 802.11 Wireless Controller Driver
HANDLE PciRoot (0x0) /Pci (0x1C, 0x1) /Pci (0x0, 0x0) /MAC (x, 0x0) Dev icePath opened 47x (G) by PS/2 Mouse Driver
HANDLE PciRoot (0x0) /Pci   (0x1C, 0x1) /Pci (0x0, 0x0) /MAC (x, 0x0) DevicePath opened 47x (T) by Platform Console Management Driver
HANDLE PciRoot (0x0) /Pci (0x1C, 0x1) /Pci (0x0, 0x0) /MAC (x  , 0x0) DevicePath opened 47x (G) by PS/2 Keyboard Driver
HANDLE PciRoot (0x0) /Pci (0x1C, 0x1) /Pci (0x0, 0x0) /MAC (x  , 0x0) DevicePath opened 47x (T) by Platform Console Management Driver
HANDLE PciRoot (0x0) /Pci (0x1C, 0x1) /Pci (0x0, 0x0) /MAC (x, 0x0) DevicePath opened 15x (T) by Simple Network Protocol Driver
HANDLE PciRoot (0x0) /Pci (0x1C, 0x1) /Pci (0x0, 0x0) /MAC (x, 0x0) Nii31 supported
HANDLE PciRoot (0x0) /Pci (0x1C, 0x1) /Pci (0x0, 0x0) /MAC (x, 0x0) Nii31 opened 1x (T) by ipxe.efi
EFIDRV PciRoot (0x0) /Pci (0x1C, 0x1) /Pci (0x0, 0x0) /MAC (x, 0x0) connecting new drivers
HANDLE PciRoot (0x0) /Pci (0x1C, 0x1) /Pci (0x0, 0x0) /MAC (x, 0x0) is a Nii31 device
EFIDRV Pci Root (0x0) /Pci (0x1C, 0x1) /Pci (0x0 , 0x0) /MAC (x, 0x0) has driver "NII"
EFIDRV PciRoot (0x0) /Pci (0x1C, 0x1) /Pci (0x0, 0x0)/MAC (x , 0x0) DRIVER_ START
EFIDRV Pci Root (0x0) /Pci (0x1C, 0x1) /Pci (0x0, 0x0) /MAC (x, 0x0) is not supported by driver "PCI": Error 0x7f10e183 (https://ipxe.org/7f10e183)
HANDLE. PciRoot (0x0) /Pci (0x1C, 0x1) /Pci (0x0 , 0x0) /MAC (x, 0x0) is a Nii31 device
```
hangs in here

```
	DBGC ( device, "HANDLE %s is a %s device\n",
```
https://github.com/ipxe/ipxe/blob/master/src/drivers/net/efi/snp.c
hangs after .. what is run after that code

```
      efi_handle_name ( device ), efi_guid_ntoa ( protocol ) );
	return 0;
}
```

thx

**Comment by mcb30** [comment]
Extend your debugging to `DEBUG=efi_driver:3,nii:3`

**Reply by kocoman1** [reply]
EFIDRV PciRoot (0x0) /Pci (0x1C, 0x1) /Pci (0x0, 0x0)/MAC (x , 0x0) DRIVER_ START
EFIDRV Pci Root (0x0) /Pci (0x1C, 0x1) /Pci (0x0, 0x0) /MAC (x, 0x0) is not supported by driver "PCI": Error 0x7f10e183 (https://ipxe.org/7f10e183)
NII NII-0000: 2b:00.0 using UNDI v0.0 at 0x1000c entry 0xbaf3213d00bbed9f impl 0x5663eed2
NII NII-0000:2b:00.0 has menory BAR 0 and no I/O BAR
NII NII-0000:2b:00.0 issuing 01:0000 ifnum 0 cpb
04dcc3d8 : 1b de da 7d 00 00 00 00-0f cf da 7d 00 00 00 00  ...}....}....
04dcc3e8 : 00 00 00 00 00 00 00 00-4d cf da 7d 00 00 00 00  .....M..}....
04dcc3f8 : 05 da da 7d 00 00 00 00-4d d5 da 7d 00 00 00 00  ...}..M..}....
04dcc408 : 26 d5 da 7d 00 00 00 00-50 48 e2 7d 00 00 00 00  &..}....] PH.J ....
 
hangs here

**Reply by mcb30** [reply]
> NII NII-0000: 2b:00.0 using UNDI v0.0 at 0x1000c entry 0xbaf3213d00bbed9f impl 0x5663eed2 

Well, this totally explains the crash.  Your UEFI firmware is providing totally garbage values in the NII data structure, including an invalid pointer to the UNDI API entry point.  When iPXE jumps to this entry point, it jumps to the garbage address provided by your UEFI firmware, and the system crashes.

**Reply by stappersg** [reply]
> Your UEFI firmware is providing totally garbage values in the NII data structure  ....  it jumps to the garbage address provided by your UEFI firmware, and the system crashes.


Options I see to get _iPXE_ working on the system with the poor UEFI implementation:
 * Use a native iPXE driver
 * Try another UEFI firmware version

**Reply by kocoman1** [reply]
Its a tianocore/coreboot.. I have to see what mnpdxe depends on and see what I can do, its impossible to debug in tianocore/edk2 easily.. plus the intel EFI code is hard to read/understand, its for BCM4360 wireless used on macbookpro and asus pcie ac wifi card. would like to get LTDP diskless working on it someday. macbook bios does work for wifi recovery of osx over internet, but haven't try loading the ipxe.efi via shell in refind yet.. but because I converted to coreboot/tianocore its a bit hassle to convert back to test, but I will do it when I get a chance
